distsql: add batching for secondary lookup joins

solongordon opened the pull request:
Lookup joins on non-covering secondary indexes were previously making a
separate primary index scan for every secondary index row. Now those
scans are grouped together into batches of up to 100 spans.

Release note: None

solongordon committed:
distsql: add batching for secondary lookup joins

Lookup joins on non-covering secondary indexes were previously making a
separate primary index scan for every secondary index row. Now those
scans are grouped together into batches of up to 100 spans.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25815)
<!-- Reviewable:end -->

solongordon commented:
Please ignore the first commit which is from a dependent PR.

solongordon commented:
Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/distsqlrun/joinreader.go, line 388 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8DxUT6XTYrrKr_-nI:-LD8DxUT6XTYrrKr_-nJ:bgqirnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L388)):*
> ```Go
> // returned, both the inputs and the output have been drained and closed, except
> // if an error is returned.
> func (jr *joinReader) indexJoinLookup(
> ```

I disentangled the index join and lookup join logic a bit because there was too much cognitive overhead. I think we could stand go further in this direction down the line.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
With regard to testing, I'd definitely like to at least see a test case that has multiple results in the outer join. There's a bunch of code here that tries to preserve ordering of the results (I think) - that doesn't seem to be tested properly if there's no test case that has more than one output.

---

Reviewed 5 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 388 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8DxUT6XTYrrKr_-nI:-LD8Hqrt822iYK73U2Q6:b-gnlf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L388)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I disentangled the index join and lookup join logic a bit because there was too much cognitive overhead. I think we could stand go further in this direction down the line.
</blockquote></details>

:thumbsup: although I wish the disentangling was in its own commit - the code movement combined with the changes makes it hard to see what's new vs what's copied.

---

*[pkg/sql/distsqlrun/joinreader.go, line 485 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8Ibib9t5nmOmnmVAv:-LD8Ibib9t5nmOmnmVAw:bf0djfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L485)):*
> ```Go
> 			// second lookup against the primary index and replace our previous
> 			// results with the primary rows.
> 			secondaryIndexRows := make([]sqlbase.EncDatumRow, len(lookupRows))
> ```

This doesn't have to be allocated every time, does it? We could allocate a slice up front and reuse it. At least put a TODO or something here so it's easy to spot if this shows up as a hot allocations spot in profiling later.

---

*[pkg/sql/distsqlrun/joinreader.go, line 494 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8J0LT-kRdtBFXuVTH:-LD8J0LUDH4s2KHVLqSp:b7y0slv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L494)):*
> ```Go
> 			}
> 			for i := range rows {
> 				lookupRows[i].row = rows[i]
> ```

Is it possible that `primaryLookup` won't always return a one to one match for all of its inputs? Do we need to check anything before replacing `lookupRows` values? I suppose this same code path was present in the last PR, but just noticing it now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-:-LD8JcRe9oAUR71x6hfr:b-u8flmc)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I would try to run all the logic tests with the lookup join flag set and see if something breaks (other than EXPLAIN (DISTSQL) statements and the like)

---

Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 456 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8JiPi2o7Dy4gxj2-5:-LD8JiPi2o7Dy4gxj2-6:b-k1ablr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L456)):*
> ```Go
> 
> 	// lookupRow represents an index key and the corresponding row.
> 	type lookupRow = struct {
> ```

Nice, TIL

---

*[pkg/sql/distsqlrun/joinreader.go, line 553 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8K2QJCMufrZ9Jqc6N:-LD8K2QJCMufrZ9Jqc6O:b-tdcg5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L553)):*
> ```Go
> 	batchSize := len(secondaryIndexRows)
> 	if batchSize == 0 {
> 		return secondaryIndexRows, nil
> ```

why `secondaryIndexRows` instead of `nil`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-:-LD8Kkld3fSkGKWDvX22:byxy3lg)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 456 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8JiPi2o7Dy4gxj2-5:-LD8WihJ7b_LaPGzzapg:b708soi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L456)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Nice, TIL
</blockquote></details>

What is the `=` doing? I know you an make a type that is local to a function, but that doesn't require the `=`. Hmm, I think that is making a type alias and I don't think it is necessary. (Not harmful, but also not necessary).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Yeah, I agree about more tests. I did just add some outer join tests with multiple results to the other PR. But really we ought to have some tests where the number of results exceeds the batch size. (I've manually tested this by setting the batch size very small and running logic tests.) I'll look into that. Also I'll see if I can try out Radu's suggestion.

---

Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 388 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8DxUT6XTYrrKr_-nI:-LDANT4z49y24t7IlW5S:b-3lusjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L388)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

:thumbsup: although I wish the disentangling was in its own commit - the code movement combined with the changes makes it hard to see what's new vs what's copied.
</blockquote></details>

Yeah, sorry about that.

---

*[pkg/sql/distsqlrun/joinreader.go, line 456 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8JiPi2o7Dy4gxj2-5:-LDANiS655Ta0Ayhf4c8:b-a9nn5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L456)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What is the `=` doing? I know you an make a type that is local to a function, but that doesn't require the `=`. Hmm, I think that is making a type alias and I don't think it is necessary. (Not harmful, but also not necessary).
</blockquote></details>

Yeah, this was just my fingers misremembering the `type` syntax. Fixed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 485 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8Ibib9t5nmOmnmVAv:-LDAPL2D4EnRohIs5amb:bw4z3ce) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L485)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This doesn't have to be allocated every time, does it? We could allocate a slice up front and reuse it. At least put a TODO or something here so it's easy to spot if this shows up as a hot allocations spot in profiling later.
</blockquote></details>

True. I added a TODO for now.

---

*[pkg/sql/distsqlrun/joinreader.go, line 494 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8J0LT-kRdtBFXuVTH:-LDAPRMH8sS5AYKwOVUs:b-ggocqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L494)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Is it possible that `primaryLookup` won't always return a one to one match for all of its inputs? Do we need to check anything before replacing `lookupRows` values? I suppose this same code path was present in the last PR, but just noticing it now.
</blockquote></details>

As far as I understand it shouldn't be possible, since each secondary index row should correspond to exactly one primary index row. But if there's some weird case I'm missing I'd love to know.

`primaryLookup` does check that it fetches exactly the same number of primary index rows as expected. Though it wouldn't catch if two primary rows somehow mapped onto the same secondary one. I could iterate over the output and assert that there are no nil values, which I think would do it, though not sure if that's necessary.

---

*[pkg/sql/distsqlrun/joinreader.go, line 553 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LD8K2QJCMufrZ9Jqc6N:-LDARQlr5x8F3H1U8sMY:b-z8mxmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1533557525ccaf5bfc4e14b3ee19eb9a3f84f733/pkg/sql/distsqlrun/joinreader.go#L553)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

why `secondaryIndexRows` instead of `nil`?
</blockquote></details>

No good reason, fixed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-:-LDARtKN9d1HncrfvX3y:ba30e8v)*
<!-- Sent from Reviewable.io -->

solongordon commented:
I ended up just making the batch size a parameter and lowering it to 2 in unit tests.

---

Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-:-LDC2V0x1zpIwbLR4SQh:bei4e12)*
<!-- Sent from Reviewable.io -->

solongordon commented:
The PR this depended on was merged and this was rebased. PTAL.

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCty449-k6ZcNmHMS8:-LDCty449-k6ZcNmHMS9:btnhuz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/distsqlrun/joinreader.go#L290)):*
> ```Go
> 
> 	var alloc sqlbase.DatumAlloc
> 	rows := make([]sqlbase.EncDatumRow, 0)
> ```

what's the point of making zero-sized slices? I'd just let it be `var rows []sqlbase.EncDatumRow`

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 232 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCuH276JfhPSxpxP5J:-LDCuH276JfhPSxpxP5K:bgv9rgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/distsqlrun/joinreader_test.go#L232)):*
> ```Go
> 
> 			// Set a lower batch size to force multiple batches.
> 			jr.batchSize = 2
> ```

[nit] could be a random value between 1 and 10, cheap way to test more cases

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 350 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCuQxiAeRoq43U-Jcd:-LDCuQxiAeRoq43U-Jce:bb73xfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/distsqlrun/joinreader_test.go#L350)):*
> ```Go
> 
> 			// Set a lower batch size to force multiple batches.
> 			jr.batchSize = 2
> ```

Same

---

*[pkg/sql/distsqlrun/processors.proto, line 261 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCuZbE6pPJGHqE-UYk:-LDCuZbE6pPJGHqE-UYl:buce3lc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/distsqlrun/processors.proto#L261)):*
> ```protobuf
>   optional Expression index_filter_expr = 5 [(gogoproto.nullable) = false];
> 
>   optional sqlbase.JoinType type = 6 [(gogoproto.nullable) = false];
> ```

Whoa, why are you changing the `type` field id to `6`? That will unnecessarily break compatibility

BTW bump the distsqlrun version and mention the change in the versions text file

---

*[pkg/sql/logictest/testdata/logic_test/lookup_join, line 395 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCvAcP4URDsHkFC2kM:-LDCvAcP4URDsHkFC2kN:b3iyvlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/logictest/testdata/logic_test/lookup_join#L395)):*
> ```
> 10 100
> 10 200
> ```

It would be good to have some tests where the distsql plan is actually distributed. Maybe create a secondary table and split that and use it (perhaps in some subquery) instead of VALUES

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/lookup_join, line 395 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCvAcP4URDsHkFC2kM:-LDDNBpZ6BM-WgdTE0wm:bnunohr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/logictest/testdata/logic_test/lookup_join#L395)):*
<details><summary><i>Previously, solongordon wrote…</i></summary><blockquote>

Yeah, makes sense. I'm going to create an issue for follow-up test enhancements since there are a couple others I'd like to make too.
</blockquote></details>

Created https://github.com/cockroachdb/cockroach/issues/25862

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815)*
<!-- Sent from Reviewable.io -->

solongordon commented:
bors r+

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 232 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCuH276JfhPSxpxP5J:-LDDRUvwEJVpZvupoa9i:b-rztdli) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/distsqlrun/joinreader_test.go#L232)):*
<details><summary><i>Previously, solongordon wrote…</i></summary><blockquote>

I like the idea though I'm worried about introducing randomness into unit tests. Could lead to flakiness.
</blockquote></details>

Many unit tests have randomness. If the test is flaky w.r.t the batch size, that's definitely a bug..

---

*[pkg/sql/distsqlrun/processors.proto, line 261 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25815#-LDCuZbE6pPJGHqE-UYk:-LDDRjpn1g8jzZF2Twgq:b-pgoya) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cae2f477a6711787b401f3497ec64f996e4f442/pkg/sql/distsqlrun/processors.proto#L261)):*
<details><summary><i>Previously, solongordon wrote…</i></summary><blockquote>

I think you must be seeing a diff from the rebase. (There was some renumbering between revisions of the PR this depended on.) `processors.proto` isn't touched by this commit.

And ohhh I didn't know about or had forgotten about the distsql version. That makes sense. I thought you were referring to the cockroach version in a previous comment. Done.
</blockquote></details>

Ah sorry I was looking at some partial diffs

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25815)*
<!-- Sent from Reviewable.io -->
